### PR TITLE
1335 rpg skilltrees implement conditionals based on design doc

### LINKED ISSUE #1335: rpg-skilltrees: implement conditionals (based on design doc)
team members: Kabir, Nikola, Chelsea

Work on implementing conditional skills using the groundwork laid for complex skills, and using dummy readers, as that will be solved in a different task.

### PR DESCRIPTION
**Pull request by nbilaver**
Everything needed to create conditionals is implemented except for readers, which is being implemented in issue #1340, this implementation has dummy versions of readers. Follows a similar structure as combined and sequential skills.

### CODE REVIEW

**Review by arya32hub — APPROVED** [review]

**Comment by nbilaver** [comment]
The implementation for conditional skills has been made. The structure is based on what we've made in the complex skills doc, and passes the tests written for it, except for executing the conditional skill, as that requires a working implementation of readers, which is being taken care of in a separate issue. Everything should be good here!

File: tests/game-state/test_player.c [hunk 1]
```diff
@@ -13,70 +13,6 @@
 #include "battle/battle_default_objects.h"
 #include "battle/battle_structs.h"
 
-/* Checks that player_quest_new properly allocates memory and initializes a player quest */
```
Comment (jacknugent1529) [inline tests/game-state/test_player.c]: Is there a reason these tests were deleted?

**Review by jacknugent1529 — COMMENTED** [review]
This all looks really good! It's well documented and design in an organized manner. I noticed there's some overlap with the complex skill PR. Does it matter which order they are merged in? Also, it looks like slight updates are needed to keep up with dev.

File: tests/game-state/test_player.c (hunk 1 above)
Comment (nbilaver) [reply tests/game-state/test_player.c]: Strange, I'm not sure why these were deleted either, after restoring them they seem to pass. I pushed a change restoring these tests.

**Comment by nbilaver** [comment]
@jacknugent1529 This was purposefully made in order to use similar code to the complex skill PR, I would recommend merging the complex skill PR first and then this one.

**Comment by nbilaver** [comment]
Merge request issues have been fixed. The code between issue #1335 and #1382 shared a lot of similar code, but developed differences between the two as development continued. The most significant difference is that for conditionals, a reader field was added to the complex_skill_t struct. Adjustments have been made so that both types of skills will be usable when merged.

File: src/skilltrees/src/complex_skills.c [hunk 2]
```diff
@@ -10,11 +10,11 @@
 #include "skilltrees/complex_skills.h"
 
 /*See complex_skills.h */
-complex_skill_t* complex_skill_new(complex_skill_type_t type, skill_t** skills, int num_skills){
+complex_skill_t* complex_skill_new(complex_skill_type_t type, skill_t** skills, int num_skills, reader_effect_t* reader){
     complex_skill_t* complex;
 
     if (num_skills == 0) {
-        fprintf(stderr, "skill_new: num_skills is invalid (given 0)\n");
+        fprintf(stderr, "complex_skill_new: max_level is invalid\n");
```
Comment (jacknugent1529) [inline src/skilltrees/src/complex_skills.c]: This error message seems to be incorrect for the context

File: tests/skilltrees/complex_skill_test.c [hunk 3]
```diff
@@ -159,72 +168,27 @@ Test(skill_test, sequential_complex_skill_execute){
...
-//     int x = complex_skill_xp_up(complex, xp);
-//     cr_assert_eq(x, 0, "Error: failed test complex_skill_xp_up_failure");
+//     complex_skill_t* complex = complex_skill_new(SEQUENTIAL, skills, 2, NULL);
+//     int x = complex_skill_level_up(complex);
+//     cr_assert_eq(x, 0, "Error: failed test complex_skill_level_up_test");
 // }
-
-
-
-/* Tests complex_skill_level_up when the return value is success */
```
Comment (jacknugent1529) [inline tests/skilltrees/complex_skill_test.c:193]: Is there a reason these are deleted?

**Review by jacknugent1529 — COMMENTED** [review]
It looks great overall, just a few small items. Also, waiting a review from @arya32hub

File: tests/skilltrees/complex_skill_test.c (hunk 3 above)
Comment (nbilaver) [reply tests/skilltrees/complex_skill_test.c:193]: The tests are currently still present in the latest version of the file, but commented out. We have marked them with a comment that both are listed to be fixed in a future sprint or skilltrees team, as the crash was giving quite a bit of trouble.

File: src/skilltrees/src/complex_skills.c (hunk 2 above)
Comment (nbilaver) [reply src/skilltrees/src/complex_skills.c]: Agreed, message has been changed to:
"complex_skill_new: subskill length of < 0 given"
Along with an adjustment to the if statement its contained in from num_skills == 0 to num_skills <= 0, which is more proper given the context.

**Review by jacknugent1529 — APPROVED** [review]
Looks good to me!

**Review by arya32hub — COMMENTED** [review]
Great to see the addition of the reader affect. Code seems solid. Haven't been able to pickup any faults so far, also good to see code is clear and easy to follow, limited to 18 characters or less per line

**Review by arya32hub — APPROVED** [review]

**Comment by KabirDubey** [comment]
Closing statement: Nik initiated the task and designed conditional skills. Implementation was evenly split in a team meeting. Nik addressed reviews.

**Comment by jacknugent1529** [comment]
Issue Score: Excellent

Comments: Excellent work! The code was well written and prepared for a PR. You addressed the comments well, although your description of the PR (the opening comment) could be a bit more detailed about what exactly conditional skills are.

**Pull request merged**